File: web/src/app/battlemap/community/community.ts
import { HttpClient, HttpErrorResponse, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { firstValueFrom } from 'rxjs';

export type ModerationStatus = 'Pending' | 'Approved' | 'Rejected';
export type MediaStatus = 'Pending' | 'Approved' | 'Rejected';

export interface CommentView {
  id: number;
  authorName: string;
  body: string;
  createdUtc: string;
  mine: boolean;
  canDelete: boolean;
}

/** A note as the signed-in person sees it: the author and editors see the newest text, everyone else the approved text. */
export interface NoteView {
  id: number;
  contactId: number;
  title: string;
  body: string;
  authorName: string;
  createdUtc: string;
  updatedUtc: string;
  status: ModerationStatus;
  /** The newest version is waiting for approval while an earlier approved one is what the public sees. */
  pendingEdit: boolean;
  mine: boolean;
  canEdit: boolean;
  commentsOpen: boolean;
  comments: CommentView[];
}

export interface NoteInput {
  title: string;
  body: string;
}

export interface VersionView {
  versionNo: number;
  title: string;
  body: string;
  editedByName: string;
  createdUtc: string;
  approved: boolean;
}

export interface IncidentMediaView {
  id: number;
  mediaId: number;
  /** The incident it belongs to. Pictures carried over from the old site may have none and be placed on the map only. */
  contactId: number | null;
  url: string;
  thumbUrl: string;
  width: number;
  height: number;
  caption: string | null;
  credit: string | null;
  dateTaken: string | null;
  lat: number | null;
  lon: number | null;
  status: MediaStatus;
  likes: number;
  likedByMe: boolean;
  mine: boolean;
  canRemove: boolean;
  /** Recorded when the picture was uploaded. Who added it is sent only with a picture's own page, not in lists. */
  byteSize: number;
  contentType: string;
  addedUtc: string;
  addedBy: string | null;
}

/** A picture taken near an incident (not one of its own), from `GET /api/contacts/{id}/nearby-media`. */
export interface NearbyPicture {
  id: number;
  contactId: number | null;
  thumbUrl: string;
  caption: string | null;
  credit: string | null;
  lat: number;
  lon: number;
  distanceMetres: number;
}

export interface LikeResult {
  likes: number;
  liked: boolean;
}

export interface HonourSummary {
  serviceNumber: string;
  name: string;
  rank: string | null;
  branch: string | null;
  birth: string | null;
  death: string | null;
  ageAtDeath: number | null;
  portraitUrl: string | null;
  /** How a roll writes the name: "White, James Mungo". */
  sortName?: string | null;
}

export interface HonourTour {
  unit: string | null;
  start: string | null;
  end: string | null;
}

export interface HonourPerson extends HonourSummary {
  birthPlace: string | null;
  birthState: string | null;
  birthCountry: string | null;
  nationalService: boolean | null;
  tours: HonourTour[];
  incidents: number[];
  tributes: number;
}

/** What to restrict the roll to; a blank or missing value is no restriction. */
export interface HonourFilters {
  service?: string;
  rank?: string;
  corps?: string;
}

/** One choice in a drop-down, and how many people it would leave. */
export interface HonourFacetOption {
  value: string;
  count: number;
}

/** The choices for each drop-down, each counted with the search and the other two choices but not its own. */
export interface HonourFacets {
  services: HonourFacetOption[];
  ranks: HonourFacetOption[];
  corps: HonourFacetOption[];
}

export interface HonourPage {
  items: HonourSummary[];
  total: number;
  page: number;
  pageSize: number;
  /** Present only when asked for. */
  facets?: HonourFacets | null;
}

export interface TributeView {
  id: number;
  authorName: string;
  message: string;
  createdUtc: string;
  mine: boolean;
  canDelete: boolean;
}

export interface TributePage {
  items: TributeView[];
  total: number;
  page: number;
  pageSize: number;
}

export interface CasualtyInput {
  serviceNumber: string | null;
  casualtyType: string;
  comment: string;
}

export interface CasualtyRow {
  id: number;
  contactId: number;
  serviceNumber: string | null;
  casualtyType: string;
  comment: string;
  submittedByName: string;
  createdUtc: string;
  handled: boolean;
}

export interface PendingNote {
  id: number;
  contactId: number;
  title: string;
  body: string;
  authorName: string;
  updatedUtc: string;
  isChange: boolean;
}

export interface PendingPicture {
  incidentMediaId: number;
  mediaId: number;
  contactId: number | null;
  url: string;
  thumbUrl: string;
  caption: string | null;
  credit: string | null;
  uploadedByName: string;
}

export interface ModerationQueue {
  notes: PendingNote[];
  pictures: PendingPicture[];
  casualties: CasualtyRow[];
}

/** What can happen to a casualty in an incident, as the API accepts it. */
export const CASUALTY_TYPES = ['Killed in action', 'Died of wounds', 'Wounded in action', 'Missing', 'Other'] as const;

/** Typed calls to the community API. Anyone may read; writing needs a signed-in member, which the server checks again. */
@Injectable({ providedIn: 'root' })
export class CommunityService {
  private readonly http = inject(HttpClient);

  // ---- notes
  notes(contactId: number): Promise<NoteView[]> {
    return firstValueFrom(this.http.get<NoteView[]>(`/api/contacts/${contactId}/notes`));
  }

  createNote(contactId: number, input: NoteInput): Promise<NoteView> {
    return firstValueFrom(this.http.post<NoteView>(`/api/contacts/${contactId}/notes`, input));
  }

  updateNote(id: number, input: NoteInput): Promise<NoteView> {
    return firstValueFrom(this.http.put<NoteView>(`/api/notes/${id}`, input));
  }

  deleteNote(id: number): Promise<void> {
    return firstValueFrom(this.http.delete<void>(`/api/notes/${id}`));
  }

  noteVersions(id: number): Promise<VersionView[]> {
    return firstValueFrom(this.http.get<VersionView[]>(`/api/notes/${id}/versions`));
  }

  moderateNote(id: number, status: ModerationStatus): Promise<NoteView> {
    return firstValueFrom(this.http.post<NoteView>(`/api/notes/${id}/status`, { status }));
  }

  setCommentsOpen(id: number, open: boolean): Promise<NoteView> {
    return firstValueFrom(this.http.post<NoteView>(`/api/notes/${id}/comments-open`, { open }));
  }

  addComment(noteId: number, body: string): Promise<NoteView> {
    return firstValueFrom(this.http.post<NoteView>(`/api/notes/${noteId}/comments`, { body }));
  }

  deleteComment(id: number): Promise<NoteView> {
    return firstValueFrom(this.http.delete<NoteView>(`/api/comments/${id}`));
  }

  // ---- pictures
  media(contactId: number): Promise<IncidentMediaView[]> {
    return firstValueFrom(this.http.get<IncidentMediaView[]>(`/api/contacts/${contactId}/media`));
  }

  /** Every approved picture with a place on the map, newest first (the server sends at most 500). */
  mediaOnMap(): Promise<IncidentMediaView[]> {
    const params = new HttpParams().set('minLat', -90).set('minLon', -180).set('maxLat', 90).set('maxLon', 180);
    return firstValueFrom(this.http.get<IncidentMediaView[]>('/api/community-media', { params }));
  }

  /** Approved pictures placed near an incident, nearest first, leaving out the incident's own. */
  nearbyMedia(contactId: number, radiusKm = 2, limit = 8): Promise<NearbyPicture[]> {
    const params = new HttpParams().set('radiusKm', radiusKm).set('limit', limit);
    return firstValueFrom(this.http.get<NearbyPicture[]>(`/api/contacts/${contactId}/nearby-media`, { params }));
  }

  /** One picture as this viewer sees it (their own like, whether they can remove it), or `null` when there is no such picture for them. */
  mediaDetail(id: number): Promise<IncidentMediaView | null> {
    return firstValueFrom(this.http.get<IncidentMediaView>(`/api/incident-media/${id}`)).catch((e) => {
      if (e instanceof HttpErrorResponse && e.status === 404) {
        return null;
      }
      throw e;
    });
  }

  addMedia(contactId: number, file: File, caption: string, credit: string, dateTaken: string): Promise<IncidentMediaView> {
    const form = new FormData();
    form.append('file', file, file.name);
    for (const [key, value] of [['caption', caption], ['credit', credit], ['dateTaken', dateTaken]] as const) {
      if (value) {
        form.append(key, value);
      }
    }
    return firstValueFrom(this.http.post<IncidentMediaView>(`/api/contacts/${contactId}/media`, form));
  }

  removeMedia(id: number): Promise<void> {
    return firstValueFrom(this.http.delete<void>(`/api/incident-media/${id}`));
  }

  toggleLike(id: number): Promise<LikeResult> {
    return firstValueFrom(this.http.post<LikeResult>(`/api/incident-media/${id}/like`, {}));
  }

  setPictureStatus(mediaId: number, status: MediaStatus): Promise<unknown> {
    return firstValueFrom(this.http.post(`/api/studio/media/${mediaId}/status`, { status }));
  }

  // ---- honour roll
  casualties(contactId: number): Promise<HonourSummary[]> {
    return firstValueFrom(this.http.get<HonourSummary[]>(`/api/contacts/${contactId}/casualties`));
  }

  submitCasualty(contactId: number, input: CasualtyInput): Promise<CasualtyRow> {
    return firstValueFrom(this.http.post<CasualtyRow>(`/api/contacts/${contactId}/casualty-submissions`, input));
  }

  /** The roll by surname, for what was typed and the chosen service, rank and corps, with the choices for those drop-downs if `facets` is set. */
  honourRoll(q: string, page = 1, pageSize = 20, filters: HonourFilters = {}, facets = false): Promise<HonourPage> {
    let params = new HttpParams().set('page', page).set('pageSize', pageSize);
    if (q.trim()) {
      params = params.set('q', q.trim());
    }
    for (const key of ['service', 'rank', 'corps'] as const) {
      const value = filters[key]?.trim();
      if (value) {
        params = params.set(key, value);
      }
    }
    if (facets) {
      params = params.set('facets', true);
    }
    return firstValueFrom(this.http.get<HonourPage>('/api/honour-roll', { params }));
  }

  person(serviceNumber: string): Promise<HonourPerson> {
    return firstValueFrom(this.http.get<HonourPerson>(`/api/honour-roll/${encodeURIComponent(serviceNumber)}`));
  }

  tributes(serviceNumber: string, page = 1): Promise<TributePage> {
    return firstValueFrom(this.http.get<TributePage>(`/api/honour-roll/${encodeURIComponent(serviceNumber)}/tributes`, { params: { page } }));
  }

  leaveTribute(serviceNumber: string, message: string): Promise<TributeView> {
    return firstValueFrom(this.http.post<TributeView>(`/api/honour-roll/${encodeURIComponent(serviceNumber)}/tributes`, { message }));
  }

  deleteTribute(id: number): Promise<void> {
    return firstValueFrom(this.http.delete<void>(`/api/tributes/${id}`));
  }

  // ---- moderation (editors)
  queue(): Promise<ModerationQueue> {
    return firstValueFrom(this.http.get<ModerationQueue>('/api/studio/moderation'));
  }

  markCasualty(id: number, handled: boolean): Promise<CasualtyRow> {
    return firstValueFrom(this.http.post<CasualtyRow>(`/api/studio/casualty-submissions/${id}/handled`, { handled }));
  }
}

File: server/src/Avw.Data/Entities/Community.cs
namespace Avw.Data.Entities;

public enum ModerationStatus
{
    Pending,
    Approved,
    Rejected,
}

/// <summary>
/// A community note about one incident. Editing never overwrites: every save is a <see cref="IncidentNoteVersion"/>, and the
/// public sees the newest <em>approved</em> version until an editor approves a newer one.
/// </summary>
public class IncidentNote
{
    public long Id { get; set; }

    /// <summary>The Elasticsearch contact id (the contacts live in Elasticsearch, so this is not a foreign key).</summary>
    public int ContactId { get; set; }

    /// <summary>The local author. Null for migrated notes whose author has not registered yet.</summary>
    public long? AuthorId { get; set; }
    public AppUser? Author { get; set; }

    /// <summary>The name shown, stored at the time so it survives an author changing or removing their account.</summary>
    public string AuthorName { get; set; } = "";

    /// <summary>Hash of the legacy author's verified email, so their old notes can be claimed when they register (see <see cref="AppUser.EmailHash"/>).</summary>
    public string? AuthorEmailHash { get; set; }

    /// <summary>The id of this note in the legacy database, so an import can be repeated without duplicating it.</summary>
    public int? LegacyId { get; set; }

    /// <summary>The state of the <em>latest</em> version. An earlier approved version may still be showing.</summary>
    public ModerationStatus Status { get; set; }

    public int LatestVersionNo { get; set; }
    public int? ApprovedVersionNo { get; set; }
    public bool CommentsOpen { get; set; } = true;
    public DateTime CreatedUtc { get; set; }
    public DateTime UpdatedUtc { get; set; }
    public long? ModeratedById { get; set; }
    public DateTime? ModeratedUtc { get; set; }

    public List<IncidentNoteVersion> Versions { get; set; } = [];
    public List<NoteComment> Comments { get; set; } = [];
}

public class IncidentNoteVersion
{
    public long Id { get; set; }
    public long NoteId { get; set; }
    public IncidentNote Note { get; set; } = null!;
    public int VersionNo { get; set; }
    public string Title { get; set; } = "";

    /// <summary>Plain text. Paragraphs are separated by blank lines; nothing is ever treated as markup.</summary>
    public string Body { get; set; } = "";

    public long? EditedById { get; set; }
    public string EditedByName { get; set; } = "";
    public DateTime CreatedUtc { get; set; }
}

public class NoteComment
{
    public long Id { get; set; }
    public long NoteId { get; set; }
    public IncidentNote Note { get; set; } = null!;
    public long? AuthorId { get; set; }
    public string AuthorName { get; set; } = "";
    public string? AuthorEmailHash { get; set; }
    public int? LegacyId { get; set; }
    public string Body { get; set; } = "";
    public DateTime CreatedUtc { get; set; }
}

/// <summary>
/// A picture from the community, attached to an incident and/or placed on the map. The file, caption, credit and approval
/// state belong to the <see cref="MediaAsset"/>.
/// </summary>
public class IncidentMedia
{
    public long Id { get; set; }

    /// <summary>The incident it belongs to. Null for a migrated picture that was only placed on the map.</summary>
    public int? ContactId { get; set; }

    public long MediaId { get; set; }
    public MediaAsset Media { get; set; } = null!;
    public long? AttachedById { get; set; }

    /// <summary>The name shown for whoever added it, kept for migrated pictures whose author has not registered.</summary>
    public string? AuthorName { get; set; }

    public string? AuthorEmailHash { get; set; }

    /// <summary>Where the picture was taken, when known.</summary>
    public double? Lat { get; set; }

    public double? Lon { get; set; }
    public DateOnly? DateTaken { get; set; }
    public DateTime CreatedUtc { get; set; }

    /// <summary>Likes given on the legacy site, whose givers have no account here yet. Added to the likes given since.</summary>
    public int LegacyLikes { get; set; }

    public int? LegacyId { get; set; }
}

public class MediaLike
{
    public long MediaId { get; set; }
    public MediaAsset Media { get; set; } = null!;
    public long UserId { get; set; }
    public AppUser User { get; set; } = null!;
    public DateTime CreatedUtc { get; set; }
}

/// <summary>A poppy: a short message left for someone on the honour roll.</summary>
public class Tribute
{
    public long Id { get; set; }

    /// <summary>The service number of the person on the nominal roll (kept in Elasticsearch).</summary>
    public string ServiceNumber { get; set; } = "";

    public long? AuthorId { get; set; }
    public string AuthorName { get; set; } = "";
    public string? AuthorEmailHash { get; set; }
    public int? LegacyId { get; set; }
    public string Message { get; set; } = "";
    public DateTime CreatedUtc { get; set; }
}

/// <summary>Someone telling us about a casualty in an incident: who, what happened to them, and where they know it from.</summary>
public class CasualtySubmission
{
    public long Id { get; set; }
    public int ContactId { get; set; }
    public string? ServiceNumber { get; set; }

    /// <summary>For example "Killed in action" or "Wounded in action".</summary>
    public string CasualtyType { get; set; } = "";

    public string Comment { get; set; } = "";
    public long? SubmittedById { get; set; }
    public string SubmittedByName { get; set; } = "";
    public DateTime CreatedUtc { get; set; }
    public bool Handled { get; set; }
    public DateTime? HandledUtc { get; set; }
    public int? LegacyId { get; set; }
}

/// <summary>Links a person on the honour roll to the incident where they became a casualty.</summary>
public class CasualtyLink
{
    public string ServiceNumber { get; set; } = "";
    public int ContactId { get; set; }
}

/// <summary>
/// One person of the nominal roll who died in service, copied from Elasticsearch so that the roll can be searched with a MySQL
/// full-text index, sorted and filtered by the database. Refreshed from Elasticsearch on a schedule; never edited by hand.
/// </summary>
public class HonourRollPerson
{
    public string ServiceNumber { get; set; } = "";

    /// <summary>"James Mungo White".</summary>
    public string Name { get; set; } = "";

    /// <summary>"White, James Mungo": how a roll writes a name.</summary>
    public string SortName { get; set; } = "";

    /// <summary><see cref="SortName"/> in lower case, so every database sorts the roll the same way.</summary>
    public string SortKey { get; set; } = "";

    public string? Rank { get; set; }

    /// <summary>The roll's "branch": the corps, for the Army.</summary>
    public string? Corps { get; set; }

    /// <summary>Army, Navy or Air Force, worked out from the rank and corps because the roll does not say.</summary>
    public string? Service { get; set; }

    public DateOnly? BirthDate { get; set; }
    public DateOnly? DeathDate { get; set; }
    public string? BirthPlace { get; set; }
    public string? BirthState { get; set; }
    public string? BirthCountry { get; set; }
    public bool? NationalService { get; set; }

    /// <summary>The tours of duty, as JSON.</summary>
    public string Tours { get; set; } = "[]";

    /// <summary>The words of the names and the service number, in lower case with a space before each: what the full-text index and the short-word search read.</summary>
    public string SearchText { get; set; } = "";
}
